pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -300,47 +297,97 @@ func (s *Service) sendMsgEvents(msg *service.StateMsg) {
 func (s *Service) startInternalListener() {
 	go func() {
 		for msg := range s.callbackChannel {
-			// TODO handle error in callback - https://github.com/hyperledger/aries-framework-go/issues/242
-			if err := s.process(msg); err != nil {
-				// TODO handle error
-				logger.Errorf(err.Error())
+			if msg.Err != nil {
+				// TODO https://github.com/hyperledger/aries-framework-go/issues/438 Cleanup/Update data on Stop Event Action
+				logger.Errorf("client action event processing failed - msgID:%s error:%s", msg.ID, msg.Err)
+				err := s.processCallbackFailure(msg.ID, msg.Err)
+				if err != nil {
+					logger.Errorf("process callback failure: %s", err)
+				}
+				continue
+			}
+
+			// TODO https://github.com/hyperledger/aries-framework-go/issues/242 - retry logic
+
+			if err := s.process(msg.ID); err != nil {
+				procErr := s.processCallbackFailure(msg.ID, err)
+				if procErr != nil {
+					logger.Errorf("process callback failure: %s", procErr)
+				}
 			}
 		}
 	}()
 }
 
+func (s *Service) processCallbackFailure(id string, processErr error) error {
+	// get the transient data
+	data, err := s.getTransientEventData(id)
+	if err != nil {
+		return fmt.Errorf("unable to fetch the event transient data: %w", err)
+	}
+
+	err = s.processFailure(data.ThreadID, data.Msg, processErr)
```

Comment:
just `return s.processFailure(data.ThreadID, data.Msg, processErr)`?  it will improve coverage.

Reply:
@sudeshrshetty  do not write code because of coverage :) sometimes it is hard to detect where errors happen and wrap them with a custom message it's cool. In that case, I do not see the usage of this wrapper so, yeah we can simplify it :)

Reply:
I agree @soluchok

Reply:
@sudeshrshetty I wanted to wrap the error here.